fastcrypto-zkp/src/bn254/poseidon/mod.rs

```diff
@@ -1,65 +1,82 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+use crate::bn254::poseidon::constants::{
+    POSEIDON_CONSTANTS_U1, POSEIDON_CONSTANTS_U10, POSEIDON_CONSTANTS_U11, POSEIDON_CONSTANTS_U12,
+    POSEIDON_CONSTANTS_U13, POSEIDON_CONSTANTS_U14, POSEIDON_CONSTANTS_U15, POSEIDON_CONSTANTS_U16,
+    POSEIDON_CONSTANTS_U2, POSEIDON_CONSTANTS_U3, POSEIDON_CONSTANTS_U4, POSEIDON_CONSTANTS_U5,
+    POSEIDON_CONSTANTS_U6, POSEIDON_CONSTANTS_U7, POSEIDON_CONSTANTS_U8, POSEIDON_CONSTANTS_U9,
+};
+use crate::FrRepr;
 use ark_bn254::Fr;
 use ark_ff::{BigInteger, PrimeField};
+use byte_slice_cast::AsByteSlice;
 use fastcrypto::error::FastCryptoError;
 use fastcrypto::error::FastCryptoError::{InputTooLong, InvalidInput};
-use once_cell::sync::OnceCell;
-use poseidon_ark::Poseidon;
+use ff::PrimeField as OtherPrimeField;
+use neptune::poseidon::HashMode::OptimizedStatic;
+use neptune::Poseidon;
 use std::cmp::Ordering;
-use std::fmt::Debug;
-use std::fmt::Formatter;
 
 /// The output of the Poseidon hash function is a field element in BN254 which is 254 bits long, so
 /// we need 32 bytes to represent it as an integer.
 pub const FIELD_ELEMENT_SIZE_IN_BYTES: usize = 32;
+mod constants;
 
-/// Wrapper struct for Poseidon hash instance.
-pub struct PoseidonWrapper {
-    instance: Poseidon,
-}
+macro_rules! define_poseidon_hash {
+    ($inputs:expr, $poseidon_constants:expr) => {{
+        let mut poseidon = Poseidon::new(&$poseidon_constants);
+        poseidon.reset();
+        for input in $inputs.iter() {
+            poseidon.input(bn254_to_fr(*input)).unwrap();
+        }
+        poseidon.hash_in_mode(OptimizedStatic);
 
-impl Debug for PoseidonWrapper {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("PoseidonWrapper").finish()
-    }
+        // Neptune returns the state element with index 1 but we want the first element.
```

Comment:
Can you elaborate a bit more? What is the rationale between the two choices?

Reply:
I've added some links to where the output is extracted from the state in neptune, poseidon-rs and circomlib's implementation. We're aligned with the latter two.

The paper doesn't state anything about which element to use. They define the hash function on n elements as a permutation, so it shouldn't matter which element you use.